Sources/SmokeAWSHttp/DataAWSHttpClientDelegate.swift

```diff
@@ -0,0 +1,176 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// You may not use this file except in compliance with the License.
+// A copy of the License is located at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+//  DataAWSHttpClientDelegate.swift
+//  SmokeAWSHttp
+//
+
+import Foundation
+import NIOHTTP1
+import SmokeHTTPClient
+import SmokeAWSCore
+import QueryCoding
+import HTTPHeadersCoding
+import HTTPPathCoding
+import LoggerAPI
+
+enum DataAWSHttpClientDelegateError: Error {
+    case invalidPayloadNotData
+}
+
+/**
+ Struct conforming to the AWSHttpClientDelegate protocol passes the body directly to
+ and from the request and response. The generic ErrorType is used to generate errors based
+ on the decoding a JSON error payload from the response body. 
+ */
+public struct DataAWSHttpClientDelegate<ErrorType: Error & Decodable>: HTTPClientDelegate {
+    private let inputQueryMapDecodingStrategy: QueryEncoder.MapEncodingStrategy?
+    
+    public init(inputQueryMapDecodingStrategy: QueryEncoder.MapEncodingStrategy? = nil) {
+        self.inputQueryMapDecodingStrategy = inputQueryMapDecodingStrategy
+    }
+    
+    public func getResponseError(responseHead: HTTPResponseHead,
+                                 responseComponents: HTTPResponseComponents) throws -> Error {
+        guard let bodyData = responseComponents.body else {
+            throw HTTPError.unknownError("Error with status '\(responseHead.status)' with empty body")
+        }
+        
+        if Log.isLogging(.debug) {
+            let asString = String(data: bodyData, encoding: .utf8) ?? ""
+            
+            Log.debug("Attempting to decode error data into JSON: \(asString)")
+        }
+        
+        // attempt to get an error of Error type by decoding the body data
+        return try JSONDecoder.awsCompatibleDecoder.decode(ErrorType.self, from: bodyData)
+    }
+    
+    public func encodeInputAndQueryString<InputType>(
+        input: InputType,
+        httpPath: String) throws -> HTTPRequestComponents
+        where InputType: HTTPRequestInputProtocol {
+            
+            let pathPostfix: String
```

Comment:
This block can be condensed down to:
`let pathPostfix = input.pathPostfix ?? ""`

Reply:
Done. Thanks!